common/logging.h

```diff
@@ -0,0 +1,137 @@
+#pragma once
+
+// Third-party includes
+#include "plog/Appenders/ColorConsoleAppender.h"
+#include "plog/Log.h"
+
+// Standard C includes
+#include <cstdlib>
+#include <cstring>
+
+// Standard C++ includes
+#include <string>
+
+namespace BoBRobotics {
+namespace Logging {
+#ifdef DEBUG
+constexpr plog::Severity DefaultLogLevel = plog::Severity::debug;
+#else
+constexpr plog::Severity DefaultLogLevel = plog::Severity::info;
+#endif
+
+//! Dummy class for initialising logging before main() is called
+class Logger
+{
+public:
+    static Logger &getInstance()
+    {
+        static Logger logger;
+        return logger;
+    }
+
+private:
+    //! Plog's default TxtFormatter is a bit verbose, so let's implement our own
+    struct Formatter
+    {
+        static plog::util::nstring header()
+        {
+            return plog::util::nstring();
+        }
+
+        static plog::util::nstring format(const plog::Record& record)
+        {
+            tm t;
+            plog::util::localtime_s(&t, &record.getTime().time);
+            plog::util::nostringstream ss;
+            ss << std::setfill(PLOG_NSTR('0')) << std::setw(2) << t.tm_hour << PLOG_NSTR(":") << std::setfill(PLOG_NSTR('0')) << std::setw(2) << t.tm_min << PLOG_NSTR(":") << std::setfill(PLOG_NSTR('0')) << std::setw(2) << t.tm_sec << PLOG_NSTR(".") << std::setfill(PLOG_NSTR('0')) << std::setw(3) << record.getTime().millitm << PLOG_NSTR(" ");
+            ss << PLOG_NSTR("[") << record.getFunc() << PLOG_NSTR("@") << record.getLine() << PLOG_NSTR("] ");
+            const auto severity = record.getSeverity();
+            if (severity != plog::Severity::info && severity != plog::Severity::verbose && severity != plog::Severity::none) {
+                ss << plog::severityToString(severity) << PLOG_NSTR(": ");
+            }
+            ss << record.getMessage() << PLOG_NSTR("\n");
+
+            return ss.str();
+        }
+    };
+
+    Logger()
+    {
+        // Get log level
+        const auto severity = getLogLevel();
+
+        // Get filename to log to, if specified in env var
+        char *filenameEnvVar = std::getenv("LOG_FILE");
+        plog::util::nstring filename;
+        if (filenameEnvVar) {
+            // String conversion is necessary for Windows (which uses wchar_t)
+            filename = filenameEnvVar;
+        }
+
+        // Do initialisation
+        plog::init(severity, filename.c_str());
+
+        /*
+         * Also print log messages to console in colour.
+         *
+         * The plog documentation reports that this should have static scope.
+         */
+        static plog::ColorConsoleAppender<Formatter> appender;
+        plog::get()->addAppender(&appender);
+    }
+
+    static constexpr auto logLevelToString(plog::Severity severity)
```

Comment:
Can you reuse the versions here https://github.com/SergiusTheBest/plog/blob/365646171e192d60fa1c2509d6a20adb5893e851/include/plog/Severity.h#L16-L49?

Reply:
I did that initially but I had a couple of issues with those versions:
1. The string names didn't seem too obvious to me (e.g. ``VERB`` for verbose)
2. If you pass in some random string it'll return a log level of none, whereas I thought it'd be better to give an error (for e.g. if you make a typo)